src/democracyworks/opentracing/compat.clj

```diff
@@ -0,0 +1,25 @@
+(ns democracyworks.opentracing.compat
+  "OpenTracing functionality compatible with [clj-newrelic].
+
+  [clj-newrelic]: https://github.com/TheClimateCorporation/clj-newrelic"
+  (:require [opentracing-clj.core :as tracing]))
+
+(defmacro defn-traced
+  "Like `defn`, but wraps the function body in an OpenTracing span that mimics
+  the default behavior of the New Relic `defn-traced` macro.
+
+  Nesting calls to `defn-traced` works similarly to the default behavior of the
+  `defn-traced` macro that it replaces: nested spans will be set as children of
+  unfinished spans.
+
+  This macro is a transitional drop-in replacement for `defn-traced`, so if you
+  need more control over the tracing behavior you should manipulate spans
+  directly."
+  [fname & body]
+  (let [span-name (str *ns* \. fname)
+        first-forms (butlast body)
+        last-form (last body)]
+    `(defn ~fname
+       ~@first-forms
+       (tracing/with-span [s# {:name ~span-name}]
+         ~last-form))))
```

Comment:
I think we need to do more heavy-duty parsing of `defn` if we want this to be a drop-in replacement. E.g. [`email-works.handlers/smtp-send`](https://github.com/democracyworks/email-works/blob/201f780dc178d25c70749364f35a45c733ea6fb1/src/email_works/handlers.clj#L62-L73).

There are a number of cases like this that I found in repos that I have checked out.
```
❯ ag --clojure --count '\(defn-traced(.|\n)*?\(\['
gov-notifications/src/election_reminders/boundary/civic_info.clj:1
voting-method-works/src/voting_method_works/handlers/reporting.clj:1
voting-method-works/src/voting_method_works/entities/voting_method_preference.clj:3
email-works/src/email_works/handlers.clj:1
election-notification-works/src/election_notification_works/entities/subscription.clj:2
partner-works/src/partner_works/handlers.clj:1
```

Though at least nothing shows up (in our code at least) for this
```
❯ ag --clojure --count '\(defn-traced(.|\n)*?(:pre|:post|:doc|:arglists)\b'
```

Reply:
I don't like this suggestion, and I'm sure there are issues w/ it, but if we need to get this out the door, what about a hack like wrapping the entire fn with tracing?

```clj
(defmacro defn-traced
  [fname & body]
  (let [span-name (str *ns* \. fname)]
    `(do
       (defn f# ~@body)
       (defn ~fname [& ~'args]
         (tracing/with-span [s# {:name ~span-name}]
           (apply f# ~'args))))))
```

Reply:
Or maybe it would be easier to snag defn parsing from [`re-frame-debux`](https://github.com/Day8/re-frame-debux/blob/5e41627f182eacd8e8951b84e004310bf66a1582/src/day8/re_frame/tracing.cljc) or [`clj-newrelic`](https://github.com/TheClimateCorporation/clj-newrelic/blob/91b8a32291efa34c0e1d456529a312af3b1cba04/src/com/climate/newrelic/trace.clj#L29)?

Reply:
@mrichards42 Yes, let me think about this a little bit more. I was hoping we could use this for 90% of cases, and add manual tracing in a few places, but if it's going to be like `defn` there shouldn't be any gotchas.

Reply:
@mrichards42 I did leverage clojure.spec a bit, but ended up avoiding monkey-patching to get better unform behavior just to avoid any issues clobbering the default spec (I don't think this is a good behavior for a library). This should cover everything in the `defn` spec, except you'll get an error when the body of the function (or a function arity) has _nothing_ in it (you can still write `nil` explicitly if you want), which should only be an issue in new code, but it's caught at macroexpansion time:

```clojure
(defn-traced i-fail
  "I am a broken function."
  [x])

;Syntax error macroexpanding clojure.core/defn at (:57:1).
;opentracing-clj.core/with-span - failed: vector? at: [:fn-tail :arity-n :bodies :params] spec: clojure.core.specs.alpha/param-list
;(opentracing-clj.core/with-span [G__5661 {:name "democracyworks.opentracing.compat.i-fail"}] [x]) - failed: vector? at: [:fn-tail :arity-1 :params] spec: :clojure.core.specs.alpha/param-list
```

Reply:
Oh, and there should have been no issues with docstrings, metadata, pre/post, etc. in the first oversimplified version, I was just missing multi-arity support. I've got a test now that should demonstrate some of that, at least.

Reply:
Looks good! We aren't going to cover cases w/ fn bodies that have multiple forms (not in a `do`), but I think that's OK as a limitation.

Reply:
Your wish is my command...